### PR TITLE
Load a duck based on duckId query param

### PR DESCRIPTION
**Pull request by alschmiedt**
1. Add the ability to load duck from the duckId query param 
2. Add this to the duck board. So when a user hits edit it will take them to their duck. 

Haven't added support for the case where there is no duckId. Currently it just goes to the default page.

### CODE REVIEW

File: appengine/pond/duck/js/duck.js [hunk 1]
```diff
@@ -193,6 +188,28 @@ Pond.Duck.init = function () {
   Pond.Duck.ignoreEditorChanges_ = false;
 };
 
+/**
+ * Update the workspace with the given xml.
+ * @param {string} xml The xml to load.
+ */
+Pond.Duck.updateWorkspace = function(xmlString) {
```
Comment (moniika) [inline appengine/pond/duck/js/duck.js]: you can replace calls to this function for:
BlocklyInterface.setCode

File: appengine/pond/duck/js/datastore.js [hunk 2]
```diff
@@ -45,6 +45,15 @@ Pond.Duck.Datastore.getAllDucks = function(onLoadCallback) {
     Pond.Duck.Datastore.makeRequest_(url, 'GET', [], onLoadCallback);
 };
 
+/**
+ * Get all the ducks for the current user.
+ * @param {!Function} onLoadCallback The function to be called when response is received.
+ */
+Pond.Duck.Datastore.getADuck = function(duckId, onLoadCallback) {
```
Comment (moniika) [inline appengine/pond/duck/js/datastore.js]: Comment needs updated.
and preferred: getADuck > getDuck

File: appengine/pond/duck/online/js/online.js [hunk 3]
```diff
@@ -46,11 +47,42 @@ Pond.Duck.Online.init = function() {
...
+    if (meta['code']['opt_xml']) {
+      Pond.Duck.changeTab(0);
+      Pond.Duck.updateWorkspace(opt_xml);
+    } else {
+      Pond.Duck.changeTab(1);
+      Blockly.utils.dom.addClass(Pond.Duck.editorTabs[0], 'tab-disabled');
+      Blockly.utils.dom.addClass(Pond.Duck.editorTabs[1], 'tab-selected');
+      Blockly.utils.dom.removeClass(Pond.Duck.editorTabs[0], 'tab-selected');
+      Pond.Duck.updateJSCode(meta['code']['js']);
+    }
```
Comment (moniika) [inline appengine/pond/duck/online/js/online.js]: use opt_xml for if condition.

I would recommend instead refactoring logic in tabHandler defined in init, the logic for updating the selected tab classes is more robust (this will not work when we add more tabs).

There is a setCode method on BlocklyInterface that works for both xml and js that you could use.

This would function the same, except that we may unecessarily be setting ignoreEditorChanges_
`var opt_xml = meta['code']['opt_xml']`
`var code = opt_xml || meta['code']['js']`
`Pond.Duck.selectTab(opt_xml ? 0 : 1); // would need to be created by refactoring tabHandler`
`Pond.Duck.ignoreEditorChanges_ = false;`
`BlocklyInterface.setCode(code)`
`Pond.Duck.ignoreEditorChanges_ = true;`

File: appengine/pond/duck/online/js/online.js [hunk 4]
```diff
@@ -46,11 +47,42 @@ Pond.Duck.Online.init = function() {
...
+    } else {
+    text = BlocklyGames.getMsg('Games_httpRequestError') + '\nStatus: '
+      + this.status;
+    document.getElementById('loading').style.display = 'none';
+    BlocklyDialogs.storageAlert(null, text);
+    }
```
Comment (moniika) [inline appengine/pond/duck/online/js/online.js]: Indentation needs fixed.

File: appengine/pond/duck/online/js/online.js [hunk 5]
```diff
@@ -168,6 +204,14 @@ Pond.Duck.Online.makeRequest = function(url, data, onLoadCallback) {
   xhr.send(data.join('&'));
 };
 
+/**
+ * Submits the form information.
+ * @param {Array<string>} requiredFieldsIds The required ids for the fields we
```
Comment (moniika) [inline appengine/pond/duck/online/js/online.js]: Array.<string> (its declared incorrectly in encodeFormElements_ jsdoc too, whoops)

File: appengine/pond_storage/get-ducks.py [hunk 6]
```diff
@@ -15,14 +15,25 @@
 limitations under the License.
 """
 
-"""Get all ducks for the current user 
+"""Given a duck id get the duck. If no duck id exists get all ducks belonging to the user.
```
Comment (moniika) [inline appengine/pond_storage/get-ducks.py]: Since we're using it for get a duck and get multiple ducks, I would like for the file to be called get.py and also to also match how the other files are named too.

Also to shorten:
Given a duck key get the duck, otherwise get all ducks belonging to the user.

File: appengine/pond_storage/get-ducks.py [hunk 7]
```diff
@@ -15,14 +15,25 @@
...
+import cgi
 import json
 from pond_storage import *
 
-duckList = get_user_ducks()
-print("Content-Type: application/json\n")
-print(json.dumps({"duckList": duckList}))
+forms = cgi.FieldStorage()
+
+if forms.has_key("duckId"):
```
Comment (moniika) [inline appengine/pond_storage/get-ducks.py]: I think its confusing to use "duckId" when the value is not the duck id, but the urlSafe version of the key. We also pass the key in the other requests and we use "key" for the same type, so I think this should be "key" instead.

File: appengine/pond_storage/get-ducks.py [hunk 8]
```diff
@@ -15,14 +15,25 @@
...
+  urlsafe_key = forms["duckId"].value
+  duck_key = ndb.Key(urlsafe=urlsafe_key)
+  duck = duck_key.get()
+  if (verify_duck(duck)):
```
Comment (moniika) [inline appengine/pond_storage/get-ducks.py]: Remove parenthesis.

File: appengine/pond/duck/board/js/board.js [hunk 9]
```diff
@@ -66,6 +66,17 @@ Pond.Duck.Board.deleteDuck = function(e) {
...
 };
 
+/**
+ * Redirect the user to a page to edit the duck.
+ * @param {Event} e The event that holds the duckId.
+ */
+Pond.Duck.Board.editDuck = function(e) {
+  document.getElementById('loading').style.display = 'table-cell';
+  var duckId = e.target.getAttribute('data-duckurl');
+  var url = window.location.origin + '/pond-duck-online?duckId=' + duckId;
```
Comment (moniika) [inline appengine/pond/duck/board/js/board.js]: I prefer using "duck" instead of "duckId" as param, but am ok to leave as is.

File: appengine/pond_storage/get-ducks.py (hunk 7 above)
Comment (alschmiedt) [reply appengine/pond_storage/get-ducks.py]: I'm not a huge fan of just key since it doesn't describe what it is a key for. What about duckKey?

File: appengine/pond/duck/js/duck.js [hunk 10]
```diff
@@ -42,6 +42,15 @@ goog.require('Pond.Visualization');
  */
 Pond.Duck.editorTabs = null;
 
+/**
+ * Object holding name of tabs.
```
Comment (moniika) [inline appengine/pond/duck/js/duck.js]: This doesn't look like names, maybe reword and mention index of tab?

File: appengine/pond/duck/js/duck.js [hunk 11]
```diff
@@ -193,6 +191,21 @@ Pond.Duck.init = function () {
   Pond.Duck.ignoreEditorChanges_ = false;
 };
 
+/**
+ * Add the correct classes to make the tab at the sepcified index look selected.
+ * @param {Array.<!Element>} tabs The list of tabs.
```
Comment (moniika) [inline appengine/pond/duck/js/duck.js]: Instead use the global variable:
`Pond.Duck.editorTabs`

File: appengine/pond_storage/get-ducks.py (hunk 7 above)
Comment (moniika) [reply appengine/pond_storage/get-ducks.py]: I like the shortness of "key", but I see your point with the descriptiveness. Another idea could be to update the urls so that when we make a request it looks like:
pond-storage/update-duck?key=1234

However, since we are using "key" in all the other files, I'd like to keep it consistent here and address it in a follow up PR to update it everywhere.

File: appengine/pond/duck/online/js/online.js [hunk 12]
```diff
@@ -46,11 +47,45 @@ Pond.Duck.Online.init = function() {
...
+
   BlocklyGames.bindClick('duckCreateButton', Pond.Duck.Online.showCreateDuckForm);
   BlocklyGames.bindClick('duckUpdateButton', Pond.Duck.Online.showUpdateDuckForm);
   BlocklyGames.bindClick('duckDeleteButton', Pond.Duck.Online.showDeleteDuckForm);
 };
 
+/**
+ * Callback for when a user loads the page with a duck id.
+ */
+Pond.Duck.Online.updateEditor= function() {
```
Comment (moniika) [inline appengine/pond/duck/online/js/online.js]: missing space before "="

File: appengine/pond/duck/js/duck.js [hunk 13]
```diff
@@ -60,21 +69,15 @@ Pond.Duck.init = function () {
...
-        if (selectedIndex == i) {
-          Blockly.utils.dom.addClass(tabs[i], 'tab-selected');
-        } else {
-          Blockly.utils.dom.removeClass(tabs[i], 'tab-selected');
-        }
-      }
+      Pond.Duck.selectTab(tabs, selectedIndex);
       Pond.Duck.changeTab(selectedIndex);
     };
   }
```
Comment (moniika) [inline appengine/pond/duck/js/duck.js]: What I meant was to factor out this whole method. It doesn't look like there's a case where you wouldn't also want to call changeTab.

If you create a method that takes selectedIndex as a parameter that updates the tab classes and calls changeTab, then you can call it in Pond.Duck.Online.updateEditor and you can bind it to the button with the correct paramter using: `BlocklyGames.bindClick(tabs[i], function() {Pond.Duck.selectTab(i)}`

Comment (moniika) [reply appengine/pond/duck/js/duck.js]: Just found out the anonymous function doesn't work, what does work is 
`Pond.Duck.selectTab.bind(this, i)`
Although I commented about it, I'm going to end up modifying things anyways when I add the new tab, so I'm cool with just leaving the refactor of selectTab as is for this PR.

File: appengine/pond/duck/js/duck.js (hunk 10 above)
Comment (alschmiedt) [reply appengine/pond/duck/js/duck.js]: done

File: appengine/pond/duck/js/duck.js (hunk 11 above)
Comment (alschmiedt) [reply appengine/pond/duck/js/duck.js]: done.

File: appengine/pond/duck/online/js/online.js (hunk 12 above)
Comment (alschmiedt) [reply appengine/pond/duck/online/js/online.js]: done

File: appengine/pond_storage/get-ducks.py (hunk 7 above)
Comment (alschmiedt) [reply appengine/pond_storage/get-ducks.py]: done
